stubs/psycopg2/psycopg2/__init__.pyi

```diff
@@ -32,6 +33,18 @@ from psycopg2._psycopg import (
     threadsafety as threadsafety,
 )
 
+_T_conn = TypeVar("_T_conn", bound=connection)
+
+@overload
+def connect(dsn: str, connection_factory: Callable[..., _T_conn], cursor_factory: None = ..., **kwargs: Any) -> _T_conn: ...
+@overload
+def connect(
+    dsn: str | None = ..., *, connection_factory: Callable[..., _T_conn], cursor_factory: None = ..., **kwargs: Any
+) -> _T_conn: ...
```

Comment:
It seems a bit awkward to annotate the `cursor_factory` param with a type that can only ever be `None` for these two overloads, but because of the `**kwargs` param, it was the most reliable way I could find so that these overloads are matched before the catch-all below. Otherwise, if one were to call, for example, `connect(connection_factory=DictConnection, cursor_factory=NamedTupleCursor)`, the return type would be `DictConnection`, which would be somewhat problematic because the actual instance _would_ be a `DictConnection`, but the instance's `cursor` method would return a `NamedTupleCursor` instance by default, while the `cursor` method's return type would indicate it is a `DictCursor` .

An alternative might be to add a couple more overloads above these two, but I thought that might be more confusing/overwhelming for a user. Here's what that might look like:

```python
@overload
def connect(
    dsn: str | None,
    connection_factory: Callable[..., connection] | None,
    cursor_factory: Callable[..., cursor] | None,
    **kwargs: Any,
) -> connection: ...
@overload
def connect(
    dsn: str | None = ...,
    connection_factory: Callable[..., connection] | None = ...,
    *,
    cursor_factory: Callable[..., cursor] | None,
    **kwargs: Any,
) -> connection: ...
@overload
def connect(dsn: str, connection_factory: Callable[..., _T_conn], **kwargs: Any) -> _T_conn: ...
@overload
def connect(dsn: str | None = ..., *, connection_factory: Callable[..., _T_conn], **kwargs: Any) -> _T_conn: ...
@overload
def connect(
    dsn: str | None = ...,
    connection_factory: Callable[..., connection] | None = ...,
    cursor_factory: Callable[..., cursor] | None = ...,
    **kwargs: Any,
) -> connection: ...
```

Thoughts welcome.